src/index.ts

```diff
@@ -18,70 +18,46 @@ export type Method = (typeof methods)[number];
 
 type TruncatedResponse = Omit<Response, 'arrayBuffer' | 'blob' | 'body' | 'clone' | 'formData' | 'json' | 'text'>;
 /** Infer request/response from content type */
-type UnwrapNonNullable<T> = T extends {
+type Unwrap<T> = T extends {
   content: { 'application/json': any };
 }
   ? T['content']['application/json']
   : T extends { content: { '*/*': any } }
   ? T['content']['*/*']
   : T;
-type Unwrap<T> = T extends NonNullable<T> | undefined ? UnwrapNonNullable<NonNullable<T>> | undefined : UnwrapNonNullable<T>;
 
 export default function createClient<T>(defaultOptions?: ClientOptions) {
   const defaultHeaders = new Headers({
     ...DEFAULT_HEADERS,
     ...(defaultOptions?.headers ?? {}),
   });
 
-  async function coreFetch<U extends keyof T, M extends keyof T[U]>(url: U, options: FetchOptions<U, M>): Promise<FetchResponse<U, M>> {
-    let { headers, body, params = {}, ...init } = options || {};
-    // URL
-    let finalURL = `${defaultOptions?.baseUrl ?? ''}${url as string}`;
-    const { path, query } = (params as BaseParams | undefined) ?? {};
-    if (path) for (const [k, v] of Object.entries(path)) finalURL = finalURL.replace(`{${k}}`, encodeURIComponent(`${v}`.trim()));
-    if (query) finalURL = `${finalURL}?${new URLSearchParams(query as any).toString()}`;
-    // headers
-    const baseHeaders = new Headers(defaultHeaders); // clone defaults (don’t overwrite!)
-    const headerOverrides = new Headers(headers);
-    for (const [k, v] of headerOverrides.entries()) {
-      if (v === undefined || v === null) baseHeaders.delete(k); // allow `undefined` | `null` to erase value
-      else baseHeaders.set(k, v);
-    }
-    // fetch!
-    const res = await fetch(finalURL, {
-      redirect: 'follow',
-      ...defaultOptions,
-      ...init,
-      headers: baseHeaders,
-      body: typeof body === 'string' ? body : JSON.stringify(body),
-    });
-    const response: TruncatedResponse = {
-      bodyUsed: res.bodyUsed,
-      headers: res.headers,
-      ok: res.ok,
-      redirected: res.redirected,
-      status: res.status,
-      statusText: res.statusText,
-      type: res.type,
-      url: res.url,
-    };
-    return res.ok ? { data: await res.json(), response } : { error: await res.json(), response };
-  }
-
   /** Gets a union of paths which have method */
   type PathsWith<M extends Method> = {
     [Path in keyof T]: T[Path] extends { [K in M]: unknown } ? Path : never;
   }[keyof T];
 
+  type FetchResponse<T> =
```

Comment:
Moving these types up was just another simplification / cleanup from before that helps TS inference.